### PR TITLE
[CARBONDATA-2291] Added datamap status and refresh command to sync data manually to datamaps

### PR DESCRIPTION
**Pull request by ravipesala**
This  PR depends on PR https://github.com/apache/carbondata/pull/2100

Currently, there is no way to synchronize datamaps manually to the fact table.  So at first added the functionality to load the datamaps manually by using command `refresh`
```
REFRESH DATAMAP <datamapName>
```
Above command syncs the data between the datamap and parent tables.

In order maintain the data consistency we require to enable or disable datamaps when data is not synchronized between fact and datamap. So added the new status called datamapstatus file under the system folder. whenever the data is out of sync between parent tables and datamap it just updates the status as disabled in datamapstatus file. After user manually refreshes the datamap it will update the status to enable.
Only the enabled datamaps are considered during query.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 
 - [x] Any backward compatibility impacted?
 
 - [x] Document update required?

 - [x] Testing done
       Tests added
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3379/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4103/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4604/

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusDetail.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * limitations under the License.
+ */
+package org.apache.carbondata.core.datamap.status;
+
+import java.io.Serializable;
+
+/**
+ * Status of each datamap
+ */
+public class DataMapStatusDetail implements Serializable {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusDetail.java]: please annotate public class

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.carbondata.core.datamap.DataMapStoreManager
+import org.apache.carbondata.core.datamap.status.DataMapStatusManager
+import org.apache.carbondata.datamap.DataMapManager
+
+/**
+ * Refresh the datamaps while sync with main table data.
+ * @param tableIdentifier
+ */
+case class CarbonDataMapRefreshCommand(dataMapName: String,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala]: move dataMapName to next line

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    tableIdentifier: Option[TableIdentifier]) extends DataCommand {
+
+  override def processData(sparkSession: SparkSession): Seq[Row] = {
+
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala]: remove empty line

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+import org.apache.carbondata.core.datamap.DataMapStoreManager;
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable;
+import org.apache.carbondata.core.metadata.schema.table.DataMapSchema;
+
+/**
+ * Maintains the status of each datamap. As per the status query will decide whether to hit datamap
+ * or not.
+ */
+public class DataMapStatusManager {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java]: If this class provide static function only, please add a private constructor

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusDetail.java (hunk 1 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusDetail.java]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala (hunk 2 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala (hunk 3 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala]: ok

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java (hunk 4 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4628/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3403/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4135/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3407/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4632/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4140/

**Comment by ravipesala** [comment]
retest this please

**Comment by ravipesala** [comment]
retest sdv please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4643/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3417/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4152/

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+import org.apache.carbondata.core.locks.CarbonLockUtil;
+import org.apache.carbondata.core.locks.ICarbonLock;
+import org.apache.carbondata.core.locks.LockUsage;
+import org.apache.carbondata.core.metadata.schema.table.DataMapSchema;
+import org.apache.carbondata.core.util.CarbonProperties;
+import org.apache.carbondata.core.util.CarbonUtil;
+
+import com.google.gson.Gson;
+
+public class DiskBasedDataMapStatusProvider implements DataMapStatusStorageProvider {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java]: add comment for this class

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+  private static final LogService LOG =
+      LogServiceFactory.getLogService(DiskBasedDataMapStatusProvider.class.getName());
+
+  private static final String DATAMAP_STATUS_FILE = "datamapstatus";
+
+  @Override public DataMapStatusDetail[] getDataMapStatusDetails() throws IOException {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java]: move @Override to previous line

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.execution.command.DataCommand
+
+import org.apache.carbondata.core.datamap.DataMapStoreManager
+import org.apache.carbondata.core.datamap.status.DataMapStatusManager
+import org.apache.carbondata.datamap.DataMapManager
+
+/**
+ * Refresh the datamaps while sync with main table data.
+ * @param tableIdentifier
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala]: remove it

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala [hunk 8]
```diff
@@ -185,6 +185,16 @@ class CarbonSpark2SqlParser extends CarbonDDLSqlParser {
         CarbonDataMapShowCommand(tableIdent)
     }
 
+  /**
+   * The syntax of show datamap is used to show datamaps on the table
+   * REFRESH DATAMAP datamapname ON TABLE tableName
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: ON TABLE is optional

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+      return new DataMapStatusDetail[0];
+    }
+
+    return dataMapStatusDetails;
+  }
+
+  @Override
+  public void updateDataMapStatus(List<DataMapSchema> dataMapSchemas, DataMapStatus dataMapStatus)
+      throws IOException {
+    ICarbonLock carbonTableStatusLock = getDataMapStatusLock();
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java]: add some comment to tell it is going to overwrite the status file

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java (hunk 5 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java]: ok

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java (hunk 6 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala (hunk 7 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRefreshCommand.scala]: ok

File: integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala (hunk 8 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/parser/CarbonSpark2SqlParser.scala]: ok

File: core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java (hunk 9 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3445/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4672/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4182/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3451/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4678/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4187/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**